Types

boryana-ilieva opened the pull request:
This PR introduces a first unobserved heterogeneity component to the model in the form of discrete mass points (types) with respect to the utility cost of full-time and part-time work.

- [ ] address #61

boryana-ilieva commented:
Points to note:
- I am not completely sure wether type_shares are implemented correctly. Current implementation is sufficient for simulation, but we need to discuss, if it is also sufficient for estimation.
- The number of types in this implementation is only semi-flexible. If we would want 3 types, changes to the model_processing module will have to be made (add two lines for each additional type). Does respy have a completely flexible solution on the number of types? Do we want to have such in soepy? It would be cool, but I could not think of an easy way to implement it so far.

Further note:
- We wanted to address test coverage in this pull request. As far as I can see, the lowest coverage is given for the solve_auxiliary module, where almost all functions are numbanized. I assume this is the issue of the low coverage rate. What is your solution for this, e.g. in respy?

boryana-ilieva commented:
@peisenha Also flexible number of types init specification is now prepared. Could you please review at this stage?
If happy, I will have to create a new regression test vault due to new form of init file and additional column "Type" in the data frame. In the current version num_types in fixed to one with share one, so that I can regression and PR test.

Will look into turning off numba for coverage check on travis this afternoon.

peisenha commented on soepy/python/pre_processing/model_processing.py:
```diff
@@ -38,7 +38,28 @@ def expand_init_dict(init_dict):
-    init_dict["DERIVED_ATTR"] = {"educ_range": educ_range, "shocks_cov": shocks_cov}
+    # Extract the number of types
+
+    if [v for k, v in init_dict["PARAMETERS"].items() if "share" in k] == 1.0:
```
The expression `v for k, v in init_dict["PARAMETERS"].items() if "share" in k` is constructed numerous times here, please create one auxiliary variable instead. Also, please add some comment here what is going on. It appears too complicated for what we want to achieve, at least at first glance :)

boryana-ilieva replied:
Creating a variable is definitely a very good idea for increasing readability. Done. In case you see a way to do the whole thing in an uneasier way in the more readable version, please share. I do tend to think in complicated ways sometimes.

peisenha commented on soepy/python/pre_processing/model_processing.py:
```diff
@@ -111,8 +132,15 @@ def group_parameters(init_dict, init_dict_flat):
 
-    init_dict_flat["theta_p"] = init_dict["PARAMETERS"]["theta_p"]
-    init_dict_flat["theta_f"] = init_dict["PARAMETERS"]["theta_f"]
+    init_dict_flat["theta_p"] = [
```
lines 135 - 140 can me merged with a for loop iterating over f and p, right?

boryana-ilieva replied:
Yes. Done

peisenha commented on soepy/test/random_init.py:
```diff
@@ -94,9 +94,29 @@ def random_init(constr=None):
+
+    # One type: ensure share equals 1
+    if num_types == 1:
+        init_dict["PARAMETERS"]["theta_p1"], init_dict["PARAMETERS"][
```
lines 103-106 are rather repeated in line 112-119, please try to merge

boryana-ilieva replied:
It seems like it, but in order to assign the integer indices 1 in case of one type and 1,2,3 etc. in case of many types, I do not see a good way of not keeping the if clause and the for loop separate.

boryana-ilieva commented on soepy/python/solve/solve_auxiliary.py:
```diff
@@ -329,14 +375,14 @@ def get_continuation_values(model_params, states_subset, indexer, emaxs):
     benefits,
 ):
-    num_choices = nonconsumption_utilities.shape[0]
+    num_choices = NUM_CHOICES
```
You are very right, we do not. Deleted.

boryana-ilieva commented on soepy/test/test_unit.py:
```diff
@@ -81,67 +83,117 @@ def test4():
+    np.testing.assert_array_equal(states_true, states[0:96, :])
 
 
 cleanup()
```
I am not sure, I get what you mean.

We want to check that the state space is exactly as we want it to be for the first 4 periods. This is formulated as a unit test and unit tests are all stored in soepy/test. I would have actually thought that this is the proper location to have it at.

peisenha replied:
Sorry, that was even more unclear than usual. Do we need the `cleanup()` call at all inside the package? Can't we just move it our testing battery instead?

boryana-ilieva replied:
I see, not needed here indeed and removed.

peisenha replied:
Then we can move the function definition for `cleanup()` outside the package as well, right?

boryana-ilieva commented on soepy/test/test_unit.py:
```diff
@@ -37,7 +38,8 @@ def test():
 
 
-def test3():
+@pytest.mark.skip(reason="adjustment due after init format revision")
```
Every small change in the init file processing has to be taken into consideration in keeping the unit test for the init file printing working. I would want to adjust it last after we have finalised the init file interface for the types. So status is: To do as soon as all other requested changes are through.

boryana-ilieva commented:
@peisenha I added a constant term as we discussed and changed the handling of the baseline type a little bit. Most importantly, if one wishes to have only one type, the init file now should simply contain the constants and no information whatsoever on type parameters. Any type information included refers to all types additional to the baseline type.
Let me know if we are happy with that. Once we are, I will allow for a variable number of types in random_init and create a new regression test vault.

peisenha commented:
Thanks, I have some clarifying questions for our call. But looks like we
will be done very quickly with this now.

On Wed, Jul 24, 2019 at 3:08 PM Boryana <notifications@github.com> wrote:

> @peisenha <https://github.com/peisenha> I added a constant term as we
> discussed and changed the handling of the baseline type a little bit. Most
> importantly, if one wishes to have only one type, the init file now should
> simply contain the constants and no information whatsoever on type
> parameters. Any type information included refers to all types additional to
> the baseline type.
> Let me know if we are happy with that. Once we are, I will allow for a
> variable number of types in random_init and create a new regression test
> vault.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSourceEconomics/soepy/pull/70?email_source=notifications&email_token=AA3QYJ4I4ZRBEXR5GDXS7ILQBBH3PA5CNFSM4H3SMUZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2WIYBY#issuecomment-514624519>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3QYJ6CAVMR2AO7D2HMS3DQBBH3PANCNFSM4H3SMUZA>
> .
>


-- 
*Philipp Eisenhauer*
Economist

Mail eisenhauer@policy-lab.org
Web www.eisenhauer.io
Repository https://github.com/peisenha

peisenha commented on development/tests/regression/run.py:
```diff
@@ -66,7 +66,11 @@ def check_vault(num_test):
 
         init_dict, expected_df = test
 
+        init_dict["PARAMETERS"]["const_p"] = init_dict["PARAMETERS"].pop("theta_p")
```
We want to be sure to clean this up before the final merge.

boryana-ilieva replied:
To be addressed after update of regression test vault

boryana-ilieva replied:
Done.

peisenha commented on soepy/python/pre_processing/model_processing.py:
```diff
@@ -38,7 +38,23 @@ def expand_init_dict(init_dict):
 
-    init_dict["DERIVED_ATTR"] = {"educ_range": educ_range, "shocks_cov": shocks_cov}
+    # Extract the number of types
+    type_shares_non_baseline = [
```
Appears `v`is not used. If so, please replace with `_`.

boryana-ilieva replied:
Done

peisenha commented on soepy/python/simulate/simulate_auxiliary.py:
```diff
@@ -77,12 +87,12 @@ def pyth_simulate(model_params, states, indexer, emaxs, covariates):
             model_params.benefits ** model_params.mu
             / model_params.mu
-            * nonconsumption_utilities[0]
+            * current_non_consumption_utilities[:, 0]
```
Very minor of course, can we use `:1` for the selection to align with Line 95?

boryana-ilieva replied:
Done

peisenha commented on soepy/test/random_init.py:
```diff
@@ -94,10 +94,25 @@ def random_init(constr=None):
+    for i in range(1, num_types):
+        # Draw random parameters
+        init_dict["PARAMETERS"]["theta_p" + str(i)], init_dict["PARAMETERS"][
+            "theta_f" + str(i)
```
Please do not use the `str` function at all and instead use proper `.format()` syntax everywhere in the code. https://pyformat.info/

boryana-ilieva replied:
Done

peisenha commented on soepy/python/shared/shared_auxiliary.py:
```diff
@@ -78,8 +80,19 @@ def calculate_log_wage_systematic(model_params, states, covariates):
+    )
+
+    for i in range(1, model_params.num_types):
+        non_consumption_utility[np.where(states[:, 5] == i)] += [
```
IMPORTANT: Doesn't this need to be a multiplication for the shifts? Please carefully revisit.

boryana-ilieva replied:
Addressed

boryana-ilieva replied:
Addressed

boryana-ilieva commented on doc/tutorial/toy_model_init_file_01_1000.yml:
```diff
@@ -1,6 +1,6 @@
 ---
```
Done

boryana-ilieva commented on doc/tutorial/toy_model_init_file_01_1000.yml:
```diff
@@ -1,6 +1,6 @@
 GENERAL:
-    num_periods: 10
+    num_periods: 40
     num_choices: 3
```
Done

boryana-ilieva commented on soepy/test/test_regression.py:
```diff
@@ -20,7 +20,13 @@ def test1(idx):
+    calculated_df = calculated_df.drop(columns=["Type"])
 
     for col in expected_df.columns.tolist():
         print(col)
```
To be addressed after updating regression test vault.

boryana-ilieva replied:
Done

peisenha commented on soepy/python/pre_processing/model_processing.py:
```diff
@@ -111,8 +127,18 @@ def group_parameters(init_dict, init_dict_flat):
+    if init_dict["DERIVED_ATTR"]["num_types"] > 1:
+        for i in ["p", "f"]:
+            init_dict_flat["theta_" + i] = [
+                v for k, v in init_dict["PARAMETERS"].items() if str("theta_" + i) in k
```
Please remove use of `str()` everywhere.

boryana-ilieva replied:
Hopefully done.

peisenha commented on soepy/test/random_init.py:
```diff
@@ -94,10 +94,26 @@ def random_init(constr=None):
+    num_types = int(np.random.choice([1, 2, 3, 4], 1))
+    # Draw shares that sum up to one
+    shares = np.random.uniform(1, 10, num_types)
+    shares /= shares.sum()
```
Please just also make the `tolist()` call in this line.
